src/self_driving/planning/mcts/monte_carlo_tree_search.cpp

```diff
@@ -52,14 +50,25 @@ void MonteCarloTreeSearch::BestAction(uint64_t simulation_number,
                             memory_constraint);
     vertex->BackPropogate(pilot_, action_map_, use_min_cost_);
   }
-  // return the best action at root
+}
+
+void MonteCarloTreeSearch::BestAction(std::vector<std::vector<pilot::ActionTreeNode>> *best_action_seq, size_t topk) {
   auto curr_node = common::ManagedPointer(root_);
   while (!curr_node->IsLeaf()) {
-    auto best_child = curr_node->BestSubtree();
-    best_action_seq->emplace_back(action_map_.at(best_child->GetCurrentAction())->GetSQLCommand(),
-                                  action_map_.at(best_child->GetCurrentAction())->GetDatabaseOid());
-    SELFDRIVING_LOG_DEBUG(action_map_.at(best_child->GetCurrentAction())->GetSQLCommand());
-    curr_node = best_child;
+    std::vector<ActionTreeNode> top;
+    std::vector<common::ManagedPointer<TreeNode>> order = curr_node->BestSubtreeOrdering();
+    NOISEPAGE_ASSERT(!order.empty(), "TreeNode should have some children");
+
+    for (size_t i = 0; i < order.size() && i < topk; i++) {
+      auto child = order[i];
+      auto action = child->GetCurrentAction();
+      auto &action_info = action_map_.at(action);
+      top.emplace_back(curr_node->GetTreeNodeId(), child->GetTreeNodeId(), action, child->GetCost(),
+                       action_info->GetDatabaseOid(), action_info->GetSQLCommand());
+    }
+
+    best_action_seq->emplace_back(std::move(top));
+    curr_node = order[0];
```

Comment:
We actually need to apply the top 1 actions (at least for `ChangeKnobAction`s) during the traversal and reverse them back. This is because the `ChangeKnobAction`s are generated based on deltas. If we don't modify the knob values, the actions will always generate commands relative to the knob values from the beginning.